### ISSUE TITLE
dask crash leads to napari crash

### ISSUE THREAD

**Issue by joshmoore** [napari/napari#1549]
## 🐛 Bug

If a dask array crashes on `.compute()`, napari crashes as well.

## To Reproduce

Steps to reproduce the behavior at the dask layer:

1. `import dask.array as da`
2. `da.from_zarr("https://s3.embassy.ebi.ac.uk/idr/zarr/v0.1/6001240.zarr/masks/0").compute()`

Caused by https://github.com/intake/filesystem_spec/issues/342

Opening an OME-Zarr with ome-zarr==0.0.11 that has masks (e.g. https://s3.embassy.ebi.ac.uk/idr/zarr/v0.1/6001240.zarr/) will do the same.

<details><summary>stack trace</summary>

```
(demo) /opt/ome-zarr-py $napari https://s3.embassy.ebi.ac.uk/idr/zarr/v0.1/6001240.zarr
WARNING: Traceback (most recent call last):
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/napari/_qt/qt_layerlist.py", line 553, in changeVisible
    self.layer.visible = True
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/napari/layers/base/base.py", line 308, in visible
    self.refresh()
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/napari/layers/base/base.py", line 633, in refresh
    self.set_view_slice()
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/napari/layers/base/base.py", line 586, in set_view_slice
    self._set_view_slice()
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/napari/layers/image/image.py", line 545, in _set_view_slice
    image = np.asarray(self.data[self.dims.indices]).transpose(order)
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/numpy/core/_asarray.py", line 85, in asarray
    return array(a, dtype, copy=False, order=order)
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/dask/array/core.py", line 1336, in __array__
    x = self.compute()
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/dask/base.py", line 166, in compute
    (result,) = compute(self, traverse=False, **kwargs)
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/dask/base.py", line 444, in compute
    results = schedule(dsk, keys, **kwargs)
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/dask/threaded.py", line 84, in get
    **kwargs
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/dask/local.py", line 486, in get_async
    raise_exception(exc, tb)
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/dask/local.py", line 316, in reraise
    raise exc
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/dask/local.py", line 222, in execute_task
    result = _execute_task(task, data)
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/dask/core.py", line 121, in _execute_task
    return func(*(_execute_task(a, cache) for a in args))
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/dask/array/core.py", line 98, in getter
    c = a[b]
  File "/Users/jamoore/opt/zarr/zarr/core.py", line 572, in __getitem__
    return self.get_basic_selection(selection, fields=fields)
  File "/Users/jamoore/opt/zarr/zarr/core.py", line 698, in get_basic_selection
    fields=fields)
  File "/Users/jamoore/opt/zarr/zarr/core.py", line 740, in _get_basic_selection_nd
    return self._get_selection(indexer=indexer, out=out, fields=fields)
  File "/Users/jamoore/opt/zarr/zarr/core.py", line 1028, in _get_selection
    drop_axes=indexer.drop_axes, fields=fields)
  File "/Users/jamoore/opt/zarr/zarr/core.py", line 1586, in _chunk_getitem
    cdata = self.chunk_store[ckey]
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/fsspec/mapping.py", line 107, in __getitem__
    result = self.fs.cat(k)
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/fsspec/asyn.py", line 204, in cat
    out = sync(self.loop, self._cat, paths, **kwargs)
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/fsspec/asyn.py", line 51, in sync
    raise exc.with_traceback(tb)
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/fsspec/asyn.py", line 35, in f
    result[0] = await future
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/fsspec/asyn.py", line 198, in _cat
    for path in paths
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/fsspec/implementations/http.py", line 144, in _cat_file
    r.raise_for_status()
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/aiohttp/client_reqrep.py", line 946, in raise_for_status
    headers=self.headers)
aiohttp.client_exceptions.ClientResponseError: 404, message='Not Found', url=URL('https://s3.embassy.ebi.ac.uk/idr/zarr/v0.1/6001240.zarr/masks/0/0.0.0.0.0')
14:18:55 WARNING Traceback (most recent call last):
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/napari/_qt/qt_layerlist.py", line 553, in changeVisible
    self.layer.visible = True
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/napari/layers/base/base.py", line 308, in visible
    self.refresh()
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/napari/layers/base/base.py", line 633, in refresh
    self.set_view_slice()
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/napari/layers/base/base.py", line 586, in set_view_slice
    self._set_view_slice()
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/napari/layers/image/image.py", line 545, in _set_view_slice
    image = np.asarray(self.data[self.dims.indices]).transpose(order)
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/numpy/core/_asarray.py", line 85, in asarray
    return array(a, dtype, copy=False, order=order)
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/dask/array/core.py", line 1336, in __array__
    x = self.compute()
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/dask/base.py", line 166, in compute
    (result,) = compute(self, traverse=False, **kwargs)
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/dask/base.py", line 444, in compute
    results = schedule(dsk, keys, **kwargs)
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/dask/threaded.py", line 84, in get
    **kwargs
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/dask/local.py", line 486, in get_async
    raise_exception(exc, tb)
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/dask/local.py", line 316, in reraise
    raise exc
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/dask/local.py", line 222, in execute_task
    result = _execute_task(task, data)
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/dask/core.py", line 121, in _execute_task
    return func(*(_execute_task(a, cache) for a in args))
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/dask/array/core.py", line 98, in getter
    c = a[b]
  File "/Users/jamoore/opt/zarr/zarr/core.py", line 572, in __getitem__
    return self.get_basic_selection(selection, fields=fields)
  File "/Users/jamoore/opt/zarr/zarr/core.py", line 698, in get_basic_selection
    fields=fields)
  File "/Users/jamoore/opt/zarr/zarr/core.py", line 740, in _get_basic_selection_nd
    return self._get_selection(indexer=indexer, out=out, fields=fields)
  File "/Users/jamoore/opt/zarr/zarr/core.py", line 1028, in _get_selection
    drop_axes=indexer.drop_axes, fields=fields)
  File "/Users/jamoore/opt/zarr/zarr/core.py", line 1586, in _chunk_getitem
    cdata = self.chunk_store[ckey]
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/fsspec/mapping.py", line 107, in __getitem__
    result = self.fs.cat(k)
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/fsspec/asyn.py", line 204, in cat
    out = sync(self.loop, self._cat, paths, **kwargs)
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/fsspec/asyn.py", line 51, in sync
    raise exc.with_traceback(tb)
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/fsspec/asyn.py", line 35, in f
    result[0] = await future
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/fsspec/asyn.py", line 198, in _cat
    for path in paths
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/fsspec/implementations/http.py", line 144, in _cat_file
    r.raise_for_status()
  File "/usr/local/anaconda3/envs/demo/lib/python3.6/site-packages/aiohttp/client_reqrep.py", line 946, in raise_for_status
    headers=self.headers)
aiohttp.client_exceptions.ClientResponseError: 404, message='Not Found', url=URL('https://s3.embassy.ebi.ac.uk/idr/zarr/v0.1/6001240.zarr/masks/0/0.0.0.0.0')
Abort trap: 6
```

</details>

## Expected behavior

Disable (or remove?) the layer if/when it fails.

## Environment

```
(demo) /opt/ome-zarr-py $naparin --info
-bash: naparin: command not found
(demo) /opt/ome-zarr-py $napari --info
napari: 0.3.6
Platform: Darwin-19.6.0-x86_64-i386-64bit
Python: 3.6.10 | packaged by conda-forge | (default, Apr 24 2020, 16:27:41)  [GCC Clang 9.0.1 ]
Qt: 5.12.5
PyQt5: 5.12.3
NumPy: 1.18.4
SciPy: 1.4.1
Dask: 2.16.0
VisPy: 0.6.4

GL version:  2.1 ATI-3.10.15
MAX_TEXTURE_SIZE: 16384

Plugins:
  - napari-plugin-engine: 0.1.5
  - ome_zarr: 0.0.11.dev0
  - omero: 0.1.1.dev5+g6d45d4c
  - svg: 0.1.3
```

**Comment by tlambert03** [comment]
Thanks @joshmoore,
Interestingly, I'm actually now thinking that this is more related to exception handling in our `event_loop.py`, and would be ~fixed~ improved by a [small change](https://github.com/napari/napari/pull/1511/files#diff-d1bc703bd5b400db6db702707c0fe4e2R75-R79) I made in #1511

Take this example, where I plant an exception on the 5th frame in `lazy_array`

```python
import numpy as np
import dask.array as da
from dask import delayed
import napari

@delayed
def lazy_array(x):
    if x == 5:
        raise ValueError("Crash!")
    return np.random.rand(256, 256)


stack = da.stack(
    [
        da.from_delayed(lazy_array(i), shape=(256, 256), dtype='float64')
        for i in range(10)
    ]
)

with napari.gui_qt():
    v = napari.view_image(stack, contrast_limits=(0, 1))
```
This will load fine, and then when the slider gets to slice 5, you get the stack trace in the console, but it _doesn't_ crash to the console, and you can keep going back to the other slices that work (seems like decent behavior).
However, if in `lazy_array` I change it to `x == 0`, then napari immediately crashes out to the console, since it doesn't even make it through the yield statement in our event loop.  If it's just crashing out to the console we want to avoid, then I a partial solution is just to wrap that yield in a try/except block as in #1511.  

also interesting: if I plant the dask exception on frame zero, then napari fails to add the layer entirely, _unless_ I use `napari.view_image(stack, contrast_limits=(0, 1), visible=False)` ... in which case it adds, but then the crash only happens when I make the layer visible.  _however_, I can't actually move the slider to see other valid frames... so that's something we should probably handle as well (to isolate dask exceptions to only the slices they affect).

**Comment by kcpevey** [comment]
The default slice has changed slightly since this was first opened. It is now range()/2. Since we gave it 10, setting `x==5` causes the crash on the initial load. It can be changed to `x==4` above in order to avoid this. 

Whether it crashes on the initial load or on a subsequent load this is still true:
>you get the stack trace in the console, but it doesn't crash to the console, and you can keep going back to the other slices that work (seems like decent behavior).

> If it's just crashing out to the console we want to avoid, then a partial solution is just to wrap that yield in a try/except block as in https://github.com/napari/napari/pull/1511.

**Is this what is desired?**

>also interesting: if I plant the dask exception on frame zero, then napari fails to add the layer entirely, unless I use napari.view_image(stack, contrast_limits=(0, 1), visible=False) ... in which case it adds, but then the crash only happens when I make the layer visible. however, I can't actually move the slider to see other valid frames... so that's something we should probably handle as well (to isolate dask exceptions to only the slices they affect).

This is no longer true. If I load it with `visible=False` then make it visible, I see the crash as described above, but if I move to a different slice, everything still functions as expected. 

In summary, the crash in the console still happens, but it doesn't affect anything in Napari itself. Unless we want to prevent the console crash from happening, I'd say this can be closed.

**Comment by jni** [comment]
I agree with your assessment @kcpevey! Others might chime in and perhaps we reopen but I'll close for now.
